<a href="https://colab.research.google.com/github/geoffwoollard/mit_intro_algo/blob/main/prob_1_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Comparison of running time.
Note values in final table are in lg scale

In [1]:
import math
import pandas as pd
from scipy.special import lambertw


In [6]:
def times():
  second = 1e6
  minute = second*60
  hour = minute*60
  day = hour*24
  month = day*30.5
  year = month*12
  century=year*100
  return([int(x) for x in [second,minute,hour,day,month,year,century]])

In [75]:
ts = times()

In [74]:
def solve_for_n(time): # in lg scale
  inv_lgn = time #2**time
  inv_sqrtn = 2*math.log2(time) # time**2
  inv_n = math.log2(time) # time
  inv_nlgn = math.log2(comp_inv_nlogn(time))
  inv_n2 = 0.5*math.log2(time) # math.sqrt(time)
  inv_n3 = 1./3.*math.log2(time) # time**(1./3.)
  inv_2n = math.log2(math.log2(time)) # math.log2(time)
  inv_nfac = math.log2(comp_inv_nfac(time))
  return([inv_lgn,inv_sqrtn,inv_n,inv_nlgn,inv_n2,inv_n3,inv_2n,inv_nfac])


In [69]:
def comp_inv_nlogn(nlogn,iterations=10):
  # https://math.stackexchange.com/a/211940
  # n_real = 1e6
  # nlogn = n_real*math.log2(n_real)
  n = nlogn # initial guess
  for _ in range(iterations):
    n = nlogn/math.log2(n)
    #print(n,n*math.log2(n),nlogn)
  return(n)

In [70]:
def comp_inv_nfac(nfac):
  # https://mathoverflow.net/a/28977
  c = 0.036534
  L = math.log((nfac+c)/math.sqrt(2*math.pi))
  W = lambertw(L/math.e).real
  inv_nfac = L/W+0.5 - 1 # -1 because n! = Gamma(n-1)
  return(inv_nfac)

In [71]:
df = pd.DataFrame({t:solve_for_n(t) for t in ts})


1000000           60000000          ...  31622400000000    3162240000000000
0    1000000.000000      6.000000e+07  ...      3.162240e+13      3.162240e+15
1         39.863137      5.167692e+01  ...      8.969202e+01      1.029797e+02
2         19.931569      2.583846e+01  ...      4.484601e+01      5.148987e+01
3         15.937239      2.141773e+01  ...      3.954074e+01      4.596733e+01
4          9.965784      1.291923e+01  ...      2.242301e+01      2.574493e+01
5          6.643856      8.612820e+00  ...      1.494867e+01      1.716329e+01
6          4.316983      4.691448e+00  ...      5.486908e+00      5.686217e+00
7          3.239365      3.480899e+00  ...      4.013123e+00      4.150310e+00

[8 rows x 7 columns]

Even with centuries of runtime, when an algorithm is $O(n!)$, it can only handle $n \sim 2^{4.15}=17$, because $18! \mu s $ is more than one century!

Even $O(n^2)$ can only handle $n=10^7$, and things get worse for $n^3$ and $2^n$

In [83]:
2**df[3162240000000000]

0             inf
1    9.999762e+30
2    3.162240e+15
3    6.879320e+13
4    5.623380e+07
5    1.467793e+05
6    5.148987e+01
7    1.775692e+01
Name: 3162240000000000, dtype: float64